cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternalsCommon.qll

```diff
@@ -0,0 +1,167 @@
+  (
+    instr instanceof WriteSideEffectInstruction or
+    instr instanceof PhiInstruction or
+    instr instanceof ReadSideEffectInstruction or
+    instr instanceof ChiInstruction or
+    instr instanceof InitializeIndirectionInstruction
+  )
+}
+
+bindingset[isGLValue]
+private CppType getThisType(Cpp::MemberFunction f, boolean isGLValue) {
+  result.hasType(f.getTypeOfThis(), isGLValue)
+}
+
+cached
+CppType getResultLanguageType(Instruction i) {
+  if i.(VariableAddressInstruction).getIRVariable() instanceof IRThisVariable
+  then
+    if i.isGLValue()
+    then result = getThisType(i.getEnclosingFunction(), true)
+    else result = getThisType(i.getEnclosingFunction(), false)
+  else result = i.getResultLanguageType()
+}
+
+CppType getLanguageType(Operand operand) { result = getResultLanguageType(operand.getDef()) }
+
+int getMaxIndirectionsForType(Type type) {
+  result = countIndirectionsForCppType(getTypeForGLValue(type))
+}
+
+private int countIndirections(Type t) {
+  result =
+    1 +
+      countIndirections([t.(Cpp::PointerType).getBaseType(), t.(Cpp::ReferenceType).getBaseType()])
+  or
+  not t instanceof Cpp::PointerType and
+  not t instanceof Cpp::ReferenceType and
+  result = 0
+}
+
+int countIndirectionsForCppType(LanguageType langType) {
+  exists(Type type | langType.hasType(type, true) |
+    result = 1 + countIndirections(type.getUnspecifiedType())
+  )
+  or
+  exists(Type type | langType.hasType(type, false) |
+    result = countIndirections(type.getUnspecifiedType())
+  )
+}
+
+class AllocationInstruction extends CallInstruction {
+  AllocationInstruction() { this.getStaticCallTarget() instanceof Cpp::AllocationFunction }
+}
+
+private predicate isSourceVariableBase(Instruction i) {
+  i instanceof VariableAddressInstruction or i instanceof AllocationInstruction
+}
+
+predicate isModifiableByCall(ArgumentOperand operand) {
+  exists(CallInstruction call, int index, CppType type |
+    type = getLanguageType(operand) and
+    call.getArgumentOperand(index) = operand and
+    if index = -1
+    then not call.getStaticCallTarget() instanceof Cpp::ConstMemberFunction
+    else not SideEffects::isConstPointerLike(any(Type t | type.hasType(t, _)))
+  )
+}
+
+cached
+private module DefCached {
+  cached
+  predicate isDef(
+    boolean certain, Instruction instr, Operand address, Instruction base, int ind, int index
+  ) {
+    exists(int ind0, CppType type, int m |
+      certain = true and
+      address =
+        [
+          instr.(StoreInstruction).getDestinationAddressOperand(),
+          instr.(InitializeParameterInstruction).getAnOperand(),
+          instr.(InitializeDynamicAllocationInstruction).getAllocationAddressOperand(),
+          instr.(UninitializedInstruction).getAnOperand()
+        ]
+    |
+      isDefImpl(address, base, ind0) and
+      type = getLanguageType(address) and
+      m = countIndirectionsForCppType(type) and
+      ind = [ind0 + 1 .. ind0 + m] and
+      index = ind - (ind0 + 1)
```

Comment:
Yeah, this is a bit strange. The case for `isUse` is easier to understand, I think. For `isUse` it's:
- `ind = [ind0 .. ind0 + m]` (which might be easier read as `ind = ind0 + [0 .. m]`?) is the set of possible number of loads that can be done on `address`. `ind0` is the number of loads already performed on `address`, and we can only perform `m` number of loads due to the type of the address (i.e., for `int*` we can do 1 load).
- `index` is useful number to work with for the rest of SSA. `index = 0` means that no additional loads need to be performed to read the value represented by this use (and so the use can be represented directly by the `operand` in the IR). Any index greater than zero means that a number of loads are required to reach the value pointed to by the operand (and these uses will be represented by `IndirectOperand` and `IndirectInstruction` in a later commit in this PR).

For `isDef` we add `1` to `ind` to account for rvalue-ness (so that a definition like `x = 42` is a definition of the value pointed to by `&x`), and we modify `index` accordingly (so that `index = 0` means that this is the value represented directly by the operand in the IR).

I hope that makes sense. if not, I'd be happy to discuss this further in our chat later.

Reply:
Yeah, that makes sense. I was just confused by the redundancy between `ind` and `index` as predicate parameters.

Reply:
Yeah, we might be able to remove the `ind` parameter altogether. I'll see if I can make that work.